pallets/poa/src/lib.rs

```diff
@@ -0,0 +1,710 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::Decode;
+/// Pallet to add and remove validators.
+use frame_support::{
+    decl_error, decl_event, decl_module, decl_storage, dispatch, ensure, fail,
+    sp_runtime::{print, SaturatedConversion},
+    traits::{Currency, Get, Imbalance, OnUnbalanced},
+};
+use frame_system::{self as system, ensure_root};
+use log::{debug, warn};
+use sp_std::prelude::Vec;
+
+extern crate alloc;
+use alloc::collections::BTreeSet;
+
+#[cfg(test)]
+mod tests;
+
+/// The pallet's configuration trait.
+pub trait Trait: system::Trait + pallet_session::Trait + pallet_authorship::Trait {
+    /// The overarching event type.
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+
+    /// Epoch length in number of slots
+    type MinEpochLength: Get<u64>;
+
+    /// Maximum no. of active validators allowed
+    type MaxActiveValidators: Get<u8>;
+
+    type Currency: Currency<Self::AccountId>;
+}
+
+// This pallet's storage items.
+decl_storage! {
+
+    trait Store for Module<T: Trait> as PoAModule {
+        /// List of active validators. Maximum allowed are `MaxActiveValidators`
+        ActiveValidators get(fn active_validators) config(): Vec<T::AccountId>;
+
+        /// Next epoch will begin after this slot number, i.e. this slot number will be the last
+        /// slot of the current epoch
+        EpochEndsAt get(fn epoch_ends_at): u64;
+
+        /// Queue of validators to become part of the active validators. Validators can be added either
+        /// to the back of the queue or front by passing a flag in the add method.
+        /// On epoch end or prematurely if forced, validators from the queue are taken out in FIFO order
+        /// and added to the active validators unless the number of active validators reaches the max allowed.
+        QueuedValidators get(fn validators_to_add): Vec<T::AccountId>;
+
+        /// List to hold validators to remove either on the end of epoch or immediately. If a candidate validator
+        /// is queued for becoming an active validator but also present in the removal list, it will
+        /// not become active as this list takes precedence of queued validators. It is emptied on each
+        /// epoch end
+        RemoveValidators get(fn validators_to_remove): Vec<T::AccountId>;
```

Comment:
Which one is planned validators? Can you describe in some detail your proposal and how do you handle short-circuited. The reason to avoid manipulating `pallet_session::Validators` directly is to avoid any accidental change that changes the validator set when it should not.

Reply:
`pallet_session::Validator` is the direct source of truth. `ActiveValidators` may be more aptly named `PlannedValidators`. If we embrace that, and stop using `PlannedValidators` as a source of truth, it can be updated during the middle of a session. 

This doesn't effect `short_circuit` logic.

I don't think `pallet_session::Validator` is intended to be manipulated directly, but rather through `SessionManager::new_session()`.

Storage space is not an issue.

Reply:
> I don't think pallet_session::Validator is intended to be manipulated directly, but rather through SessionManager::new_session().

Not confident about this. Maybe there are other approved ways to edit the validator set.

Reply:
I follow you. But I am still not sure pallet_session::Validator will not be modified without us noticing. Its my lack of understanding.

Reply:
Like i said before, I need to know if validator set changed in `should_end_session`